libcxx/include/__chrono/utc_clock.h

```diff
@@ -0,0 +1,155 @@
+    // - To store the sum, allowing a binary search on the data.
+    // - Avoid acquiring a mutex.
+    // The disadvantage are:
+    // - A slightly larger code size.
+    //
+    // There are two optimization directions
+    // - hard-code the database and do a linear search for future entries. This
+    //   search can start at the back, and should probably contain very few
+    //   entries. (Adding leap seconds is quite rare and new release of libc++
+    //   can add the new entries; they are announced half a year before they are
+    //   added.)
+    // - During parsing the leap seconds store an additional database in the
+    //   dylib with the list of the sum of the leap seconds. In that case there
+    //   can be a private function __get_utc_to_sys_table that returns the
+    //   table.
+    //
+    // Note for to_sys there are no optimizations to be done; it uses
+    // get_leap_second_info. The function get_leap_second_info could benefit
+    // from optimizations as described above; again both options apply.
+
+    // Both UTC and the system clock use the same epoch. The Standard
+    // specifies from 1970-01-01 even when UTC starts at
+    // 1972-01-01 00:00:10 TAI. So when the sys_time is before epoch we can be
+    // sure there both clocks return the same value.
+
+    const tzdb& __tzdb = chrono::get_tzdb();
+    _Rp __result{__time.time_since_epoch()};
+    for (const auto& __leap_second : __tzdb.leap_seconds) {
+      if (__leap_second > __time)
+        return __result;
+
+      __result += __leap_second.value();
+    }
+    return __result;
+  }
+};
+
+struct leap_second_info {
+  bool is_leap_second;
+  seconds elapsed;
+};
+
+template <class _Duration>
+[[nodiscard]] _LIBCPP_HIDE_FROM_ABI leap_second_info get_leap_second_info(const utc_time<_Duration>& __time) {
+  const tzdb& __tzdb = chrono::get_tzdb();
+  if (__tzdb.leap_seconds.empty()) [[unlikely]]
+    return {false, chrono::seconds{0}};
+
+  sys_seconds __sys{chrono::floor<seconds>(__time).time_since_epoch()};
+  seconds __elapsed{0};
+  for (const auto& __leap_second : __tzdb.leap_seconds) {
+    if (__sys == __leap_second.date() + __elapsed)
+      return {__leap_second.value() > 0s, // only positive leap seconds are considered leap seconds
+              __elapsed + __leap_second.value()};
+
+    if (__sys < __leap_second.date() + __elapsed)
+      return {false, __elapsed};
+
+    __elapsed += __leap_second.value();
+  }
+
+  return {false, __elapsed};
+}
+
+template <class _Duration>
+[[nodiscard]] _LIBCPP_HIDE_FROM_ABI sys_time<common_type_t<_Duration, seconds>>
+utc_clock::to_sys(const utc_time<_Duration>& __time) {
+  using _Dp               = common_type_t<_Duration, seconds>;
+  leap_second_info __info = get_leap_second_info(__time);
+
+  sys_time<common_type_t<_Duration, seconds>> __result{__time.time_since_epoch() - __info.elapsed};
+  if (__info.is_leap_second)
+    return chrono::floor<seconds>(__result) + chrono::seconds{1} - _Dp{1};
```

Comment:
Floating-point concerns aside, I don't understand how this works. My (naive) understanding is that `_Dp` can be smaller than seconds, for example if `common_type_t<_Duration, seconds>` is `chrono::milliseconds`. In that case, you'd be flooring the result, then adding 1 second and subtracting one millisecond. So basically you'd be adding `0.999` seconds to the result. I don't understand why that is correct. If that's a bug, let's add a test!

Reply:
The relevant language is [[time.clock.utc.members]/2](https://eel.is/c++draft/time.clock#utc.members-2): "[If `u` represents a time during a leap second insertion,] the last representable value of sys_time prior to the insertion of the leap second is returned."

Briefly, I think the part you might be missing is that `leap_second_info::elapsed` is being subtracted before adding `seconds{1} - _Dp{1}`, and `elapsed` is incremented as soon as the insertion begins. So in this branch, when `is_leap_second` is `true`, `floor<seconds>(__result)` is one second _before_ the leap second insertion began.

To illustrate, suppose that we're examining a day with a positive leap second. To make things easier, assume that the UTC epoch is midnight of that day, this is the first leap second ever, and the _Dp` ratio is `ratio<1,4>`. Then:

| UTC time | `__time` | `__info` | `__result` | `floor<seconds>(__result) + seconds{1} - _Dp{1}` | Returned system time
| - | - | - | - | - | - |
| 23:59:59.75 | 345,599 | {false, 0s} | 345,599 | N/A | 23:59:59.75 |
| 23:59:60.00 |  345,600 | {true,1s} | 345,596 | 345,599 (=345,596+4-1) | 23:59:59.75 |
| 23:59:60.25 | 345,601 | {true,1s} | 345,597 | 345,599 | 23:59:59.75 |
| 23:59:60.50 | 345,602 | {true,1s} | 345,598 | 345,599 | 23:59:59.75 |
| 23:59:60.75 | 345,603 | {true,1s} | 345,599 | 345,599 | 23:59:59.75 |
| 00:00:00.00 | 345,604 | {false,1s} | 345,600 | N/A | 00:00:00.00 |

Reply:
@MattStephanson I read your comment, but I'm not sure what the issue is. "the last representable value of sys_time prior to the insertion of the leap second is returned." Looking at the `floor<seconds>(__result) + seconds{1} - _Dp{1}` and "Returned system time" columns these are the values returned. Could you show the values you would have expected?

Reply:
@mordante I think the returned values are correct. I was trying to explain to @ldionne why the code is correct (except, as I said earlier, for floating point).